### PR TITLE
application: serial_lte_modem: Add #XWIFIPOS command

### PR DESCRIPTION
**Pull request by junqingzou**
Add #XWIFIPOS to get location by nRF Cloud Wi-Fi location service. 
Rename "positioning" to "location" service following nRF Cloud. 
Rename CONFIG_SLM_CELL_POS to CONFIG_SLM_LOCATION. 
Correct some doc errors.

### CODE REVIEW

File: applications/serial_lte_modem/doc/GNSS_AT_commands.rst [hunk 1]
```diff
@@ -849,25 +849,19 @@ Response syntax
...
 Example
 ~~~~~~~
 
 ::
 
-  AT#XAGPS?
+  AT#XCELLPOS?
 
-  #XAGPS: 1,1
+  #XCELLPOS: 1,1
```
Comment (junqingzou) [inline applications/serial_lte_modem/doc/GNSS_AT_commands.rst:865]: Mistake, will be fixed

File: applications/serial_lte_modem/doc/GNSS_AT_commands.rst [hunk 2]
```diff
@@ -893,3 +887,135 @@ Example
...
+The command accepts ``<ssidX>`` and ``<macX>`` of up to 5 access points.
+
+Unsolicited notification
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+::
+
+   #XLOCATION: <type>,<latitude>,<longitude>,<uncertainty>
+
+* The ``<type>`` value indicates in which mode the cellular location service is running:
```
Comment (junqingzou) [inline applications/serial_lte_modem/doc/GNSS_AT_commands.rst:935]: Mistake, will be fixed

File: applications/serial_lte_modem/doc/GNSS_AT_commands.rst [hunk 3]
```diff
@@ -849,25 +849,19 @@ Response syntax
...
-* The ``<gnss_service>`` value is an integer.
-  When it returns the value of ``1``, it means that GNSS is supported in ``%XSYSTEMMODE`` and is activated in ``+CFUN``.
-
-.. note::
-   CELLPOS does not require the GNSS service in modem.
+   #XCELLPOS: <cellpos_status>
 
 * The ``<cellpos_status>`` value is an integer.
-  When it returns the value of ``1``, it means that CELLPOS is started.
+  When it returns the value of ``1``, it means that cellular location service is started.
```
Comment (wiba-nordic) [inline applications/serial_lte_modem/doc/GNSS_AT_commands.rst:855]:
```suggestion
  When it returns the value of ``1``, it means that the cellular location service is started.
```

**Review by wiba-nordic — CHANGES_REQUESTED** [review]
We're adding something new, so there should also be a changelog entry.

**Review by wiba-nordic — APPROVED** [review]

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 4]
```diff
@@ -756,7 +779,7 @@ static void on_cloud_evt_cell_pos_data_received(const struct nrf_cloud_data *con
 			LOG_INF("TTFF %ds", (int)k_uptime_delta(&ttft_start)/1000);
 			ttft_start = 0;
 		}
-		rsp_send("\r\n#XCELLPOS: %d,%lf,%lf,%d\r\n",
+		rsp_send("\r\n#XLOCATION: %d,%lf,%lf,%d\r\n",
```
Comment (VTPeltoketo) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:782]: Should cell based location have the old unsolicited notification XCELLPOS and only wifi based location uses XLOCATION? In this way we can keep backward compatibility of old AT-functionality.

Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:782]: That is possible, then it would become
`#XCELLPOS: ....`
`#XWIFIPOS: ....`
respectively

Comment (VTPeltoketo) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:782]: Yes, I would prefer this way :)

**Comment by junqingzou** [comment]
Dear reviewers, please move PR forward. Thanks!

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 5]
```diff
@@ -1354,6 +1387,105 @@ int handle_at_cellpos(enum at_cmd_type cmd_type)
...
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		err = at_params_unsigned_short_get(&at_param_list, index, &op);
+		if (err < 0) {
+			return err;
+		}
+
+		if (op == WIFIPOS_START && nrf_cloud_ready && run_type == RUN_TYPE_NONE) {
+			if (count < (2 + NRF_CLOUD_LOCATION_WIFI_AP_CNT_MIN * 2)) {
+				LOG_ERR("Too few APs %d, minimal %d", (count - 2) / 2,
```
Comment (jayteemo) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:1415]:
```suggestion
				LOG_ERR("Too few APs %d, minimum %d", (count - 2) / 2,
```

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 6]
```diff
@@ -1354,6 +1387,105 @@ int handle_at_cellpos(enum at_cmd_type cmd_type)
...
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		err = at_params_unsigned_short_get(&at_param_list, index, &op);
+		if (err < 0) {
+			return err;
+		}
+
+		if (op == WIFIPOS_START && nrf_cloud_ready && run_type == RUN_TYPE_NONE) {
+			if (count < (2 + NRF_CLOUD_LOCATION_WIFI_AP_CNT_MIN * 2)) {
```
Comment (jayteemo) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:1414]: a comment here describing why this math needs to be done would be helpful

**Review by jayteemo — APPROVED** [review]

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c (hunk 6 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:1414]: Currently SLM only send SSID and MAC address of each AP to nRF Cloud, which makes the NRF_CLOUD_LOCATION_WIFI_AP_CNT_MIN * 2. The first 2 is index 0 for AT command and index 1 for op code.

Will rephrase

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 7]
```diff
@@ -1354,6 +1387,105 @@ int handle_at_cellpos(enum at_cmd_type cmd_type)
...
+				if (err != WIFI_MAC_ADDR_LEN) {
+					LOG_ERR("Parse MAC address %d error", wifi_data.cnt);
+					return -EINVAL;
+				}
+				ap->mac[0] = mac_addr[0];
+				ap->mac[1] = mac_addr[1];
+				ap->mac[2] = mac_addr[2];
+				ap->mac[3] = mac_addr[3];
+				ap->mac[4] = mac_addr[4];
+				ap->mac[5] = mac_addr[5];
```
[the remark this answers is not in the document]
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:1449]: Cannot, assigning uint32_t to uint8_t.

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 8]
```diff
@@ -410,6 +422,8 @@ static void cell_pos_req_wk(struct k_work *work)
 			err = nrf_cloud_location_request(&cell_data, NULL, true, NULL);
 			if (err) {
 				LOG_ERR("Failed to request MCELL, error: %d", err);
+				rsp_send("\r\n#XCELLPOS: %d\r\n", err);
+				run_type = RUN_TYPE_NONE;
```
Comment (VTPeltoketo) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:426]: What kind of error codes are replied to the AT-command? They are not listed in the documentation.

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 9]
```diff
@@ -422,6 +436,21 @@ static void cell_pos_req_wk(struct k_work *work)
...
 
+static void wifi_pos_req_wk(struct k_work *work)
+{
+	int err;
+
+	/* Request location using Wi-Fi access point info */
+	err = nrf_cloud_location_request(NULL, &wifi_data, true, NULL);
+	if (err) {
+		LOG_ERR("Failed to request Wi-Fi location, error: %d", err);
+		rsp_send("\r\n#XWIFIPOS: %d\r\n", err);
```
Comment (VTPeltoketo) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:447]: What kind of error codes are replied to the AT-command? They are not listed in the documentation.

**Review by VTPeltoketo — COMMENTED** [review]
Ok,. However, I cannot verify the functionality of the WIFI part due to missing HW.

In erroneous cases, the AT-reply message is unclear.

**Comment by junqingzou** [comment]
> Ok,. However, I cannot verify the functionality of the WIFI part due to missing HW.
> 
> In erroneous cases, the AT-reply message is unclear.

@VTPeltoketo Thanks for the CI testing.
Wi-Fi scanning is not part of this PR, and not part of nRF9160, either. This PR is for evaluating of nRF Cloud Wi-Fi Location, while customer could obtain AP info in any way, by or not by nRF7002.

SLM does not document standard libc error code. There is a message [here](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/serial_lte_modem/doc/slm_extending.html#error-handling)

**Review by VTPeltoketo — APPROVED** [review]

**Pull request merged**